Issue 2058: create keep alive entry after on-line initialization

progier389 opened the pull request:
Bug description:
Keep alive entry is not created on target master after on line initialization, and its RUVelement stays empty until a direct update is issued on that master 

Fix description:
The patch allows a consumer (configured as a master) to create (if it did not exist before) the consumer's keep alive entry. It creates it at the end of a replication session at a time we are sure the changelog exists and will not be reset. It allows a consumer to have RUVelement with csn in the RUV at the first incoming replication session.

That is basically lkrispen's proposal with an associated pytest testcase 

Reviewed by:

Platforms tested: F32

relates: #[2058](https://github.com/389ds/389-ds-base/issues/2058)

droideck commented on dirsrvtests/tests/tickets/ticket2058_test.py:
```diff
@@ -0,0 +1,131 @@
+import subprocess
+import time
+import ldap
+from lib389._constants import *
```
Please, try to avoid mass imports. It makes code less readable and less explicit (which is python coding best practices).

Firstyear replied:
@droideck This is probably from the create test template, I think @mreynolds389 has done a cleanup from it recently....

droideck commented on dirsrvtests/tests/tickets/ticket2058_test.py:
```diff
@@ -0,0 +1,131 @@
+if DEBUGGING:
+    logging.getLogger(__name__).setLevel(logging.DEBUG)
+else:
+    logging.getLogger(__name__).setLevel(logging.INFO)
```
You can skip this part. It's in lib389 now :)

progier389 replied:
It was also created by create_test template, 
 Anyway fixing mreynolds389's remark about avoiding to create new testcase in ticket directory will take care of that ...

Firstyear replied:
Mark recently did a change to the template create that removes this too :)

droideck commented on dirsrvtests/tests/tickets/ticket2058_test.py:
```diff
@@ -0,0 +1,131 @@
+def check_single_keep_alive(topo,instance,dn,expected):
+    # check that keep alive entry does not exists on master1
+    try:
+        ent = topo.ins[instance].getEntry(dn, ldap.SCOPE_BASE, "(objectclass=ldapsubentry)", ['nsUniqueId', 'modifierTimestamp'])
```
Please, use direct python-ldap calls if you need something specific.
https://www.python-ldap.org/en/python-ldap-3.3.0/reference/ldap.html#ldap.LDAPObject.search_s

We plan to deprecate 'Entry' object some time soon. 

P.S. we also can add an interface for getting 'keep alive' entries to lib389 and use it from there.

Firstyear replied:
Well, if we have a keepalive we want to be able to check, we probably could add a lib389 dsldapobject type for it to make this a bit clearer...

progier389 replied:
I get DeprecationWarning when using search_s  ...
DsLdapObject seems the way to go these days. that said creating a class for keepalive entries seems an overkill,
 as for now we only check their presence (and I am not sure we should really expose them in lib389 API)
My feeling is that we should rather provide a generic way to get a set of entries ...

Firstyear replied:
Yeah, it's been asked about before, to have a way to get a single entry by dn, or to do an objectClass=* style search. Maybe I should let go of my ideals and create it :)

progier389 replied:
Getting a single object sound fine but beware to the filter:
   subentries are hidden if (objectclass=subentry) is not an explicit part of the filter
   so the filter should rather be:   (|(objectclass=*)(objectclass=subentry)) 

[progier@linux-3 slapd]$ ldapsearch -LLL -Y EXTERNAL -H ldapi://$S1 -b "cn=repl keep alive 1,dc=example,dc=com"  "(objectclass=*)" dn 2>/dev/null
[progier@linux-3 slapd]$ ldapsearch -LLL -Y EXTERNAL -H ldapi://$S1 -b "cn=repl keep alive 1,dc=example,dc=com"  "(|(objectclass=*)(objectclass=ldapsubentry))" dn 2>/dev/null
dn: cn=repl keep alive 1,dc=example,dc=com

[progier@linux-3 slapd]$

progier389 replied:
Sorry, In the previous message I meant objectclass=ldapsubentry rather than objectclass=subentry !

droideck commented on dirsrvtests/tests/tickets/ticket2058_test.py:
```diff
@@ -0,0 +1,131 @@
+def test_ticket2058(topo):
+    """Checks that replication keep alive entry are created after on line initialisations
+       (without the fix, only the entry from master1 is created)
+
```
Also, add additional field here - `:id: UUID`.
You can get UUID like this: `python3 -c "import uuid; print(uuid.uuid4())"`

progier389 replied:
Thanks !
I now understand that these :id:  are random uuid.  Ok I will gerate them using  uuid -m -v 4

droideck commented on dirsrvtests/tests/tickets/ticket2058_test.py:
```diff
@@ -0,0 +1,131 @@
+        f"--bind-method SIMPLE --bind-dn {REPLICATION_MANAGER_DN} --bind-passwd {REPLICATION_MANAGER_PASSWORD}  agmt")
+    run(f"dsconf standalone2 repl-agmt create --suffix {SUFFIX} "
+        f"--conn-protocol LDAP --host {master1.host} --port {master1.port} "
+        f"--bind-method SIMPLE --bind-dn {REPLICATION_MANAGER_DN} --bind-passwd {REPLICATION_MANAGER_PASSWORD}  agmt")
```
It's more efficient to use direct lib389 calls. It makes the code easier to debug and modify.
I see that you want to create a topology and init it later.
Check the code from this test case:
https://github.com/389ds/389-ds-base/blob/master/dirsrvtests/tests/suites/replication/regression_test.py#L195
Or the code from the link Mark posted:
https://github.com/389ds/389-ds-base/blob/master/src/lib389/lib389/cli_conf/replication.py

And feel free to ping me or other team members.
lib389 documentation is a bit poor now...

Firstyear replied:
Yeah, I was going to say, don't you want a topology fixture here? There are some examples in the replication suite ....

droideck replied:
I think `create_topology` function is built the way so it does init during the creation implicitly.
We can write another topology fixture which doesn't do that but I think the case is very rare... So it's better to setup replication here as Pierre has done (but make it with lib389 tools as I mentioned above).

progier389 replied:
Finally I found an easier method: using 2 master topology then reinitialize the replicas from an ldif without any replication metadata.  I already tried this method without success, because generating the ldif is not so easy (as exporting ldif without the replication data still export them) but seeing one of the testcase in regression_test have shown me how to easily filter the ldif file. So exporting the db then filter ldif to remove nsUniqueId and keepalive entries bring the replica state
 to the same state than the CLI (i.e: replication role and agreements are enabled but replication is stalled because master
  has different generationIds)

droideck reviewed (changes requested):
Otherwise, it's a very good code! Thanks!
And yes, as @mreynolds389  mentioned, please, put the code to `suites`.
Probably, this path will be the best place:
`dirsrvtests/tests/suites/replication/regression_test.py`

Firstyear commented on ldap/servers/plugins/replication/repl_extop.c:
```diff
@@ -1175,6 +1175,10 @@ multimaster_extop_EndNSDS50ReplicationRequest(Slapi_PBlock *pb)
+                    /* now that the changelog is open and started, we can alos cretae the
+                     * keep alive entry without risk that db and cl will not match
+                     */
+                    replica_subentry_check(replica_get_root(r), replica_get_rid(r));
```
When is this call normally made? What's the problem that this resolves? It would be good to understand what this fix is changing.

progier389 replied:
I will try to clarify what lkrispenz and tbordaz explained in the issue:
  When the keep alive entry feature was originally added, tbordaz added a call to replica_subentry_check
    in replica_enable_replication to insure that keep alive entry is already there.
 (Another place it is called is on supplier side when initializing a replica agremeent)

In ticket [48445](https://pagure.io/389-ds-base/issue/48445) the creation of a keep alive entry after online init was 
 removed because of race condition with the changelog handling but ludwig
And some time later this issue was created because keep alive entry is not created on the master (that are init by another master) and lkrispenz proposed to add it when ending the online init session once changelog is fully started

If your question is what keep alive entries are for and what is their impact, then here is the answer:
  As I understand it, these entries exist only to trigger a change that get replicated over the topology:
  Their main purpose is to generate records in the changelog and they are updated from time to time by fractional
  replication  to insure that at least a change must be replicated by FR after a great number of not replicated changes are
  found in the changelog. The interest is that the fractional RUV get then updated so less changes need to be walked
    in the changelog when searching for the first change to send)

So creating a keep alive entry has little impact as far as the replication is fully started (including the changelog) 
    (As tbordaz mentioned, the main difference is that the master will appear in the RUV ))

IMHO the main interests of this fix are:
   - it is a good exercise for me to discover the 389 the open development process
   - more importantly, it will avoid having people getting puzzled by the fact that a master does not appear in the RUV 
     and spending time to understand why (I suspect that it what happened to lkrispenz and why he opened the issue -;) )

Firstyear replied:
That makes sense, and sounds good to me. I wonder if this (really good) explanation could be in the code comments somewhere? That way it's preserved for future-us to remember. Thank you!

progier389 replied:
Hum, That seems a good idea: 
 the part about explaining the keep alive entries role could be set in comment just before the definition of the
  replica_subentry_* functions that creates, checks and updates the keep alive entries.

progier389 committed:
ticket 2058: Add keep alive entry after on-line initialization - second version

Bug description:
Keep alive entry is not created on target master after on line initialization,
and its RUVelement stays empty until a direct update is issued on that master

Fix description:
The patch allows a consumer (configured as a master) to create (if it did not
exist before) the consumer's keep alive entry. It creates it at the end of a
replication session at a time we are sure the changelog exists and will not
be reset. It allows a consumer to have RUVelement with csn in the RUV at the
first incoming replication session.

That is basically lkrispen's proposal with an associated pytest testcase

Second version changes:
   - moved the testcase to suites/replication/regression_test.py
   - set up the topology from a 2 master topology then
    reinitialized the replicas from an ldif without replication metadata
    rather than using the cli.
   - search for keepalive entries using search_s instead of getEntry
   - add a comment about keep alive entries purpose

last commit:
   - wait that ruv are in sync before checking keep alive entries

Reviewed by: droideck, Firstyear

Platforms tested: F32

relates: #2058

droideck commented on dirsrvtests/tests/suites/replication/regression_test.py:
```diff
@@ -924,6 +948,109 @@ def test_moving_entry_make_online_init_fail(topology_m2):
+               excludeSuffixes=None, repl_data=False,
+               outputfile=ldif_file, encrypt=False)
+    # Remove replication metadata that are still in the ldif
+    _remove_replication_data(ldif_file)
```
One out of three runs has failed here because it didn't find the `ldif_file`.
Maybe, it's worth to set a small `time.wait()` after `db2ldif` task.

progier389 replied:
Strange:   As far as I remember the off line export is supposed to be synchronous.
  (That is not an ldap task that need monitoring) but it is a ns-slapd instance in db2ldif mode)  
So I am not sure that adding a delay will help.
I am running the test in loop to try to reproduce the failure

progier389 replied:
So far I did not yet reproduced the issue but I hit another one: step 5 is sometime failing because the keep alive entry was not yet replicated +=> Must wait that master1 is in sync with master2 before verifying the entries.
Restarted the tests after adding a wait_for_ruv(m2,m1)

progier389 replied:
Test run in loop without error since 2 hours. I also double checked and indeed off-line export is synchronous.
So I do not know what happened in your environment (At this point: either the file was not created or it has been removed, both scenario let me think that something unusual happened in your environment (maybe the instance was been cleared by another test or some cleanup task in background ... ))

droideck replied:
> Test run in loop without error since 2 hours. I also double checked and indeed off-line export is synchronous.
> So I do not know what happened in your environment (At this point: either the file was not created or it has been removed, both scenario let me think that something unusual happened in your environment (maybe the instance was been cleared by another test or some cleanup task in background ... ))

Looks like this.
I test on an isolated environment (freshly created libvirt machine) but still, something could happen there.

If the test will fail on our CI, we will see. The failure was purely test-related (and, probably, env-related) so no worries :)

You have my ack! And I think we can merge.

droideck reviewed (approved):
Besides the small issue, you have my ack!
Really great job with the Python and lib389!
